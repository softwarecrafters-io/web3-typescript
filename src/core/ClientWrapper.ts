import Web3 from 'web3';
import { from, map, Observable } from 'rxjs';

export class ClientWrapper {
	constructor(private client: Web3) {}

	static create(urlNode: string) {
		const provider = new Web3.providers.HttpProvider(urlNode);
		const client = new Web3(provider);
		return new ClientWrapper(client);
	}

	getAccounts() {
		return from(this.client.eth.getAccounts());
	}

	getBalance(walletAddress: string) {
		return from(this.client.eth.getBalance(walletAddress, 'latest'));
	}

	sendTransaction(fromAddress: string, toAddress: string, value: string) {
		return from(this.client.eth.sendTransaction({ from: fromAddress, to: toAddress, value }));
	}
}
