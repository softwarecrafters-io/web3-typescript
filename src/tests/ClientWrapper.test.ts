import { ClientWrapper } from '../core/ClientWrapper';
import { firstValueFrom } from 'rxjs';

describe('The client should', () => {
	const client = ClientWrapper.create('http://localhost:8545');

	it('gets accounts for a given node', async () => {
		const action = await firstValueFrom(client.getAccounts());

		expect(action).toMatchSnapshot();
	});

	it('gets balance for a given account', async () => {
		const account = (await firstValueFrom(client.getAccounts()))[0];

		const action = await firstValueFrom(client.getBalance(account));

		expect(action).toBe('0');
	});

	it('does not send transactions between accounts without enough funds', async (done) => {
		const accounts = await firstValueFrom(client.getAccounts());

		const action = client.sendTransaction(accounts[0], accounts[1], '1');

		action.subscribe({
			error: (err) => {
				expect(err).toBe('');
				done();
			},
		});
	});
});
