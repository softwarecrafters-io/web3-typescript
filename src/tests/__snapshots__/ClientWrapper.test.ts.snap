// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The client should gets accounts 1`] = `
Array [
  "0x73fF4e4b5C9Ade8004b8a85fb77057e3969b81a2",
  "0x2e4d185063B247E0Ad56a0d02eD6fC47BCF6DAF1",
  "0x533DFFc1408214557C7495b34a8EB008db86A2b9",
  "0xa2CE749f103e70f104DCff3522384C19F152910e",
]
`;

exports[`The client should gets accounts 2`] = `
Array [
  "0x73fF4e4b5C9Ade8004b8a85fb77057e3969b81a2",
  "0x2e4d185063B247E0Ad56a0d02eD6fC47BCF6DAF1",
  "0x533DFFc1408214557C7495b34a8EB008db86A2b9",
  "0xa2CE749f103e70f104DCff3522384C19F152910e",
]
`;

exports[`The client should gets accounts for a given node 1`] = `
Array [
  "0x73fF4e4b5C9Ade8004b8a85fb77057e3969b81a2",
  "0x2e4d185063B247E0Ad56a0d02eD6fC47BCF6DAF1",
  "0x533DFFc1408214557C7495b34a8EB008db86A2b9",
  "0xa2CE749f103e70f104DCff3522384C19F152910e",
]
`;

exports[`The client should transaction two numbers 1`] = `
Array [
  "0x73fF4e4b5C9Ade8004b8a85fb77057e3969b81a2",
  "0x2e4d185063B247E0Ad56a0d02eD6fC47BCF6DAF1",
  "0x533DFFc1408214557C7495b34a8EB008db86A2b9",
  "0xa2CE749f103e70f104DCff3522384C19F152910e",
]
`;
